kernel/src/hil/hmac.rs

```diff
@@ -0,0 +1,45 @@
+//! Interface for HMAC
+
+use crate::mem::AppSlice;
+use crate::mem::Shared;
+use crate::returncode::ReturnCode;
+
+#[derive(PartialEq)]
+pub enum HMACHash {
```

Comment:
The approach we took on Titan [h1b/hil/digest.rs](https://github.com/google/tock-on-titan/blob/master/h1b/src/hil/digest.rs) was to have a digest engine, which you could configure to do SHA or SHA-based HMAC. The benefit of this is that you don't have to potentially add two levels of virtualization. With this API, you need to potentially virtualize on top of the hardware (in case someone wants to use it for something besides HMAC, and so needs to enqueue). This will be important if we want to enable basic SHA256 mode for the OT IP block.

Reply:
I'm not clear what you mean here.

Reply:
So, the issue is, what if two capsules/structures want to perform HMAC at the same time? You have to virtualize HMAC and enqueue them.

But what if the same piece of hardware, in addition to HMAC, does standard SHA hashes? I now have two different APIs, both of which need queueing. Some clients may want to use both pieces of functionality, so invoke both SHA and HMAC. This means they'll either need some kind of combined trait, or at least two separate client callbacks. But the semantics of the APIs will be such that you can call HMAC, have it enqueued, then also call SHA and have it enqueued.

Underneath, the two ways to solve this are either to have a complex and custom virtualization implementation (which handles the different kinds of outstanding operations), or to do two levels of virtualization: one virtualizes the narrow trait (SHA or HMAC) on top of a virtualizer for the underlying hardware implementation, which provides both traits.

Does that make more sense?

Reply:
Yeah, that makes more sense.

So I'm thinking that I could pretty much just rename the current HMAC implementation to digest. Add a new virtual_hmac capsule (pretty much like the current one) and that should be it. I'll give that a go

Reply:
I pushed an update that add two levels of virtulisation.